video/enhanced_samples_generator.h

```diff
@@ -14,37 +16,46 @@
 class EnhancedSamplesGenerator : public WaveformGenerator {
 	public:
 		EnhancedSamplesGenerator(std::shared_ptr<AudioSample> sample);
+		~EnhancedSamplesGenerator();
 
 		void setFrequency(int value);
 		void setSampleRate(int value);
 		int getSample();
+		void setSample(std::shared_ptr<AudioSample> sample);
 
 		int getDuration(uint16_t frequency);
 
 		void seekTo(uint32_t position);
 	private:
-		std::weak_ptr<AudioSample> _sample;
+		// std::weak_ptr<AudioSample> _sample;
+		std::shared_ptr<AudioSample> _sample;
 
-		uint32_t		index;				// Current index inside the current sample block
-		uint32_t		blockIndex;			// Current index into the sample data blocks
-		int32_t			repeatCount = 0;	// Sample count when repeating
+		uint32_t	index;				// Current index inside the current sample block
+		uint32_t	blockIndex;			// Current index into the sample data blocks
+		int32_t		repeatCount;		// Sample count when repeating
 		// TODO consider whether repeatStart and repeatLength may need to be here
 		// which would allow for per-channel repeat settings
 
-		int				frequency = 0;
-		int				previousSample = 0;
-		int				currentSample = 0;
-		double			samplesPerGet = 1.0;
-		double			fractionalSampleOffset = 0.0;
+		std::atomic<int>		frequency;
+		std::atomic<int>		previousSample;
+		std::atomic<int>		currentSample;
+		std::atomic<double>		samplesPerGet;
+		std::atomic<double>		fractionalSampleOffset;
+		std::mutex				sampleMutex;
 
 		double calculateSamplerate(uint16_t frequency);
 		int8_t getNextSample();
 };
 
 EnhancedSamplesGenerator::EnhancedSamplesGenerator(std::shared_ptr<AudioSample> sample)
-	: _sample(sample)
+	: _sample(sample), repeatCount(0), index(0), blockIndex(0), frequency(0), previousSample(0), currentSample(0), samplesPerGet(1.0), fractionalSampleOffset(0.0)
 {}
 
+EnhancedSamplesGenerator::~EnhancedSamplesGenerator() {
+	sampleMutex.lock();
```

Comment:
belt & braces, ensure we're not disposing of this waveform generator when the mutex is locked from elsewhere